synapse/lib/stormlib/macro.py

```diff
@@ -0,0 +1,184 @@
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.storm as s_storm
+import synapse.lib.stormtypes as s_stormtypes
+
+macro_set_descr = '''
+Set a macro definition in the cortex.
+
+Example:
+    macro.add foobar ${ [+#foo] }
+'''
+
+macro_del_descr = '''
+Remove a macro definition from the cortex.
+'''
+
+macro_get_descr = '''
+Display the storm query for a macro in the cortex.
+'''
+
+macro_list_descr = '''
+List the macros set on the cortex.
+'''
+
+stormcmds = [
+    {
+        'name': 'macro.del',
+        'descr': macro_del_descr,
+        'cmdargs': (
+            ('name', {'help': 'The name of the macro to set.'}),
+        ),
+        'storm': '''
+            $lib.macro.del($cmdopts.name)
+            $lib.print('Removed macro: {name}', name=$cmdopts.name)
+        ''',
+    },
+    {
+        'name': 'macro.set',
+        'descr': macro_set_descr,
+        'cmdargs': (
+            ('name', {'help': 'The name of the macro to set.'}),
+            ('storm', {'help': 'The storm command string or embedded query to set.'}),
+        ),
+        'storm': '''
+            $lib.macro.set($cmdopts.name, $cmdopts.storm)
+            $lib.print('Set macro: {name}', name=$cmdopts.name)
+        ''',
+    },
+    {
+        'name': 'macro.get',
+        'descr': macro_get_descr,
+        'cmdargs': (
+            ('name', {'help': 'The name of the macro to display.'}),
+        ),
+        'storm': '''
+            $mdef = $lib.macro.get($cmdopts.name)
+            if $mdef {
+                $lib.print($mdef.storm)
+            } else {
+                $lib.print('Macro not found: {name}', name=$cmdopts.name)
+            }
+        ''',
+    },
+    {
+        'name': 'macro.list',
+        'descr': macro_list_descr,
+        'storm': '''
+            $count = $(0)
+            for ($name, $mdef) in $lib.macro.list() {
+                $user = $lib.auth.users.get($mdef.user)
+                $lib.print('{name} (owner: {user})', name=$name.ljust(20), user=$user.name)
+                $count = $($count + 1)
+            }
+            $lib.print('{count} macros found', count=$count)
+        ''',
+    },
+]
+
+class MacroExecCmd(s_storm.Cmd):
+    # FIXME Give a description of use.
+    '''
+    Execute a named macro.
+    '''
+
+    name = 'macro.exec'
+
+    def getArgParser(self):
+        pars = s_storm.Cmd.getArgParser(self)
+        pars.add_argument('name', help='The name of the macro to execute')
+        return pars
+
+    async def execStormCmd(self, runt, genr):
+
+        if not self.runtsafe:
+            mesg = 'macro.exec does not support per-node invocation'
```

Comment:
really? This seems a major limitation

Reply:
doing this in a non-runtsafe fashion means we have to get the macro defintion for each node inline, potentially.

this prevents doing 
```
macro.exec $macroName
```